### ISSUE TITLE
Deprecate `torch.squeeze()` with unspecified `dim`

### ISSUE THREAD

**Issue by YodaEmbedding** [pytorch/pytorch#142852]
### Proposed new feature or change:

Deprecate calls unspecified `dim` in calls to `torch.squeeze()`, and/or issue a single runtime warning.

### Rationale

`squeeze()` without an `dim` argument is the cause of many subtle bugs. Example:

```python
x = images[:batch_size]

B, C, H, W = x.shape

# Incorrect attempt to remove "grayscale" channel dimension.
gray = x.squeeze()
```

This causes hard-to-debug issues if:

- `B == 1`  (batch size is 1)
- `H == 1`  (height is 1)
- `W == 1`  (width is 1)

The above should have been written:

```python
gray = x.squeeze(1)
```

### Examples in the wild

Many users of this library are not "experts" at code, and frequently produce and publish programs containing this bug.

[YOLOX (10k stars)](https://github.com/Megvii-BaseDetection/YOLOX/blob/d872c71bf63e1906ef7b7bb5a9d7a529c7a59e6a/yolox/utils/boxes.py#L49):
```python
        conf_mask = (image_pred[:, 4] * class_conf.squeeze() >= conf_thre).squeeze()
```

The correct code would have been:
```python
        conf_mask = image_pred[:, 4] * class_conf.squeeze(1) >= conf_thre
```

The above bug actually inspired the creation of this proposal. See also: https://github.com/numpy/numpy/issues/27968

---

P.S. On a similar note, `torch.zeros(42).squeeze(0)` should also issue a runtime warning. NumPy correctly raises a full-fledged error in this case.

**Comment by mikaylagawarecki** [comment]
Due to the complexity in deprecating `torch.squeeze()` without the `dim` argument (and the number of downstream users cited), this might be something better addressed via adding a lint rule to torchfix cc @kit1980
